File: src/day08/main.rs
#![allow(unused_imports)]
use std::{iter::Map, ops::RangeInclusive};

use itertools::FoldWhile::{Continue, Done};
use itertools::Itertools;
use shared::puzzle_input;

fn north_from(grid: &[Vec<u32>], x: usize, y: usize) -> Vec<u32> {
    grid[0..y].iter().rev().map(|row| row[x]).collect_vec()
}
fn south_from(grid: &[Vec<u32>], x: usize, y: usize) -> Vec<u32> {
    grid[y + 1..].iter().map(|row| row[x]).collect_vec()
}
fn east_from(grid: &[Vec<u32>], x: usize, y: usize) -> Vec<u32> {
    grid[y][x + 1..].iter().cloned().collect_vec()
}
fn west_from(grid: &[Vec<u32>], x: usize, y: usize) -> Vec<u32> {
    grid[y][0..x].iter().rev().cloned().collect_vec()
}
fn sight_lines(grid: &[Vec<u32>], x: usize, y: usize) -> Vec<Vec<u32>> {
    vec![
        north_from(grid, x, y),
        south_from(grid, x, y),
        east_from(grid, x, y),
        west_from(grid, x, y),
    ]
}

fn visible_from_outside(h: u32, sight_lines: &[Vec<u32>]) -> bool {
    sight_lines
        .iter()
        .any(|trees| trees.iter().all(|&h2| h2 < h))
}

fn p2_score(h: u32, sight_lines: &[Vec<u32>]) -> usize {
    sight_lines
        .iter()
        .map(|line| {
            let x = line.iter().fold_while(0, |count, h2| {
                if h2 >= &h {
                    Done(count + 1)
                } else {
                    Continue(count + 1)
                }
            });
            match x {
                Continue(x) => x,
                Done(x) => x,
            }
        })
        .product::<usize>()
}

pub fn main() {
    let input = puzzle_input!();
    let heights = input
        .split('\n')
        .map(|line| line.chars().map(|c| c.to_digit(10).unwrap()).collect_vec())
        .collect_vec();
    let data = (0..heights.len()).map(|y| {
        let row = &heights[y];
        (0..row.len())
            .map(|x| (row[x], sight_lines(&heights, x, y)))
            .collect_vec()
    });

    let visible_from_outside = data.clone().map(|row| {
        row.iter()
            .map(|(h, lines)| visible_from_outside(*h, lines))
            .collect_vec()
    });
    let p1: usize = visible_from_outside
        .map(|row| row.iter().filter(|x| **x).count())
        .sum();
    let p2: usize = data
        .map(|row| {
            row.iter()
                .map(|(h, lines)| p2_score(*h, lines))
                .max()
                .unwrap()
        })
        .max()
        .unwrap();

    println!("part 1: {}\npart 2: {}", p1, p2);
}

File: src/day08/main.exs
defmodule Day08 do
  def sight_line(xs, ys, heights) do
    for x <- xs, y <- ys do
      heights
      |> Map.get({x, y})
    end
  end

  def sight_lines({x, y}, heights, grid_size) do
    [
      sight_line(x..x, (y - 1)..0//-1, heights),
      sight_line(x..x, (y + 1)..(grid_size - 1)//1, heights),
      sight_line((x - 1)..0//-1, y..y, heights),
      sight_line((x + 1)..(grid_size - 1)//1, y..y, heights)
    ]
  end

  def sight_distance(hs, h) do
    case Enum.find_index(hs, &(&1 >= h)) do
      nil -> Enum.count(hs)
      i -> i + 1
    end
  end
end

{input, grid_size} = {File.read!("puzzle_input/day08"), 99}

heights =
  input
  |> String.trim_trailing()
  |> String.split(~r/\R+/, trim: True)
  |> Stream.with_index()
  |> Stream.flat_map(fn {line, y} ->
    String.graphemes(line)
    |> Stream.with_index()
    |> Enum.map(fn {s, x} ->
      {h, _} = Integer.parse(s)
      {{x, y}, h}
    end)
  end)
  |> Map.new()

sight_lines =
  Map.new(heights, fn {xy, h} ->
    {xy, {Day08.sight_lines(xy, heights, grid_size), h}}
  end)

p1 =
  Enum.count(sight_lines, fn {_, {lines, h}} ->
    Enum.any?(lines, fn sight_line ->
      Enum.all?(sight_line, fn x -> x < h end)
    end)
  end)

IO.puts("part 1: #{p1}")

p2 =
  Enum.map(sight_lines, fn {_, {lines, h}} ->
    Enum.map(lines, fn line ->
      Day08.sight_distance(line, h)
    end)
    |> Enum.product()
  end)
  |> Enum.max()

IO.puts("part 2: #{p2}")
